File: ios/Cove/SettingsScreen/WipeDataPinView.swift
//
//  WipeDataPinView.swift
//  Cove
//
//  Created by Praveen Perera on 12/17/24.
//

import SwiftUI

private enum PinState {
    case new
    case confirm(String)
}

struct WipeDataPinView: View {
    /// args
    var onComplete: (String) -> Void
    var backAction: () -> Void

    /// private
    @State private var pinState: PinState = .new

    var body: some View {
        Group {
            switch pinState {
            case .new:
                NumberPadPinView(
                    title: "Enter Wipe Data PIN",
                    isPinCorrect: { _ in true },
                    showPin: false,
                    backAction: backAction,
                    onUnlock: { enteredPin in
                        withAnimation {
                            pinState = .confirm(enteredPin)
                        }
                    }
                )
            case let .confirm(pinToConfirm):
                NumberPadPinView(
                    title: "Confirm Wipe Data PIN",
                    isPinCorrect: { $0 == pinToConfirm },
                    showPin: false,
                    backAction: backAction,
                    onUnlock: onComplete
                )
            }
        }
    }
}

#Preview {
    WipeDataPinView(onComplete: { _ in }, backAction: {})
        .environment(AuthManager())
}

File: android/app/src/main/java/org/bitcoinppl/cove/flows/SettingsFlow/WipeDataPinView.kt
package org.bitcoinppl.cove.flows.SettingsFlow

import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.res.stringResource
import org.bitcoinppl.cove.R
import org.bitcoinppl.cove.views.NumberPadPinView

private sealed class WipeDataPinState {
    data object New : WipeDataPinState()

    data class Confirm(
        val pinToConfirm: String,
    ) : WipeDataPinState()
}

@Composable
fun WipeDataPinView(
    onComplete: (String) -> Unit,
    backAction: () -> Unit,
) {
    var pinState: WipeDataPinState by remember { mutableStateOf(WipeDataPinState.New) }

    when (val state = pinState) {
        is WipeDataPinState.New -> {
            NumberPadPinView(
                title = stringResource(R.string.settings_pin_enter_wipe_data),
                isPinCorrect = { true },
                backAction = backAction,
                onUnlock = { enteredPin ->
                    pinState = WipeDataPinState.Confirm(enteredPin)
                },
            )
        }

        is WipeDataPinState.Confirm -> {
            NumberPadPinView(
                title = stringResource(R.string.settings_pin_confirm_wipe_data),
                isPinCorrect = { it == state.pinToConfirm },
                backAction = backAction,
                onUnlock = onComplete,
            )
        }
    }
}
